File: scripts/verify_go_version.py
import subprocess
import sys
import os
import re

def error(message, code = 1):
    # print(message, file=sys.stderr)
    sys.exit(code)

def usage():
    return f"Usage: {sys.argv[0]} [target min go version]"

def get_system_go_version():
    which = subprocess.run(
        ["which", "go"],
        capture_output=True,
    )

    if which.returncode:
        error("go: command not found")

    output = subprocess.run(
        ["go", "version"],
        capture_output=True,
        text=True
    )

    pattern = r"\bgo(\d+)\.(\d+)\.(\d+)\b"
    match = re.search(pattern, output.stdout)
    if match:
        major, minor, patch = map(int, match.groups())
        return (major, minor, patch)
    error("System go version doesn't match the go versionning system (goXX.YY.ZZ) !")

def parse_go_semver(entry):
    pattern = r"^go(\d+)\.(\d+)\.(\d+)$"
    match = re.match(pattern, entry)
    if match:
        major, minor, patch = map(int, match.groups())
        return (major, minor, patch)
    error(f"\"{entry}\" doesn't match the go versionning system (goXX.YY.ZZ) !")

def main():
    if len(sys.argv) != 2:
        error(usage(), 1)

    min_version = parse_go_semver(sys.argv[1])
    cur_version = get_system_go_version()

    for i in range(3):
        if cur_version[i] < min_version[i]:
            error(f"Installed go version go{'.'.join((str(i) for i in cur_version))} is older than minimum required version go{'.'.join((str(i) for i in min_version))}")
    return 0

if __name__ == "__main__":
    sys.exit(main())

File: scripts/verify_go_version.rb
require 'open3'

def error(message, code = 1)
  # STDERR.puts(message)
  exit(code)
end

def usage
  "Usage: #{File.basename($PROGRAM_NAME)} [target min go version]"
end

def get_system_go_version
  go_path, status = Open3.capture2('which go')
  error('go: command not found') unless status.success?

  output, status = Open3.capture2('go version')
  error('Failed to retrieve Go version') unless status.success?

  pattern = /\bgo(\d+)\.(\d+)\.(\d+)\b/
  match = output.match(pattern)
  if match
    major, minor, patch = match.captures.map(&:to_i)
    return [major, minor, patch]
  end

  error("System Go version doesn't match the Go versioning system (goXX.YY.ZZ)!")
end

def parse_go_semver(entry)
  pattern = /^go(\d+)\.(\d+)\.(\d+)$/
  match = entry.match(pattern)
  if match
    major, minor, patch = match.captures.map(&:to_i)
    return [major, minor, patch]
  end

  error("\"#{entry}\" doesn't match the Go versioning system (goXX.YY.ZZ)!")
end

def main
  error(usage, 1) unless ARGV.size == 1

  min_version = parse_go_semver(ARGV[0])
  cur_version = get_system_go_version

  3.times do |i|
    if cur_version[i] < min_version[i]
      error("Installed Go version go#{cur_version.join('.')} is older than the minimum required version go#{min_version.join('.')}")
    end
  end

  0
end

exit(main)
